### ISSUE TITLE
How to bin new samples with splits generated from training?

### ISSUE THREAD

**Issue by GabrielSGoncalves** [guillermo-navas-palencia/optbinning#26]
Hi,

First of all, congrats for the library!

I'm trying to use the bins generated and use it to bin data in production.
Here is my example:

```python
from optbinning import OptimalBinning

x = df["var1"].to_numpy()
y = df["target"].to_numpy()

optb = OptimalBinning(name='var1')
optb.fit(x, y)
optb.splits
```
Then I get an array with the splits (for a numerical variable).
That's great, but if I want to categorize 'var1' for one row of new data (not part of training), what's the best way to do it?

One way to do it using `pandas.cut()`, but I don't know if there is a better way to do it using `optbinning`.
```python
pd.cut(df_new["var1"],  bins=optb.splits)
```
Also, how can I deal with categorical variables, as some bins are more than one category merged into an array (eg: `['cat1', 'cat3', 'catn']`).

Att

Gabriel

**Comment by guillermo-navas-palencia** [comment]
Hi Gabriel,

Thanks for your comment!

OptimalBinning, and other binning algorithms, includes the method ``transform`` to transform/categorize new data. The resulting data is transformed to event rate or Weight of Evidence (WoE). As an example, if you check the binning table, ``x_new = 12.3``, will be transformed to the event rate or WoE corresponding to the bin [12, 13). This method can handle either numerical and categorical variables. Try:
```python
x_woe = optb.transform(df_new["var1"], metric="woe")
x_event_rate = optb.transform(df_new["var1"], metric="event_rate")
```
To check the correctness of the transformation, you can use ``pandas`` as follows:

```python
pd.Series(x_woe).value_counts()
```
Perhaps, you just want the corresponding array of indices, then using ``numpy.digitize`` might be sufficient:

```python
numpy.digitize(df_new["var1"], bins=optb.splits)
```
Note however, that this will not work for categorical variables. Therefore, I would recommend the ``transform`` method. In case you rather prefer the corresponding indices, the method ``transform`` could be adapted with a new metric option, say ``metric="indices"``, which should be straightforward to implement.

Let me know whether I misunderstood your comment, and you are looking for something different.

Thanks,
Guillermo

**Comment by GabrielSGoncalves** [comment]
Hi Guilhermo,

Thanks for the fast answer!

The `transform` method with `metrics='indices'` is exactly what I need.
I would be extremely handy for creating ScoreCards and to use in production.

Would you be willing to implement it? 
If you need a hand for this task I'll be glad to help.

I really appreciate your attention,

Gabriel

**Comment by guillermo-navas-palencia** [comment]
Hi Gabriel,

I can manage to implement it this week. A problem arises for categorical variables: if df_new incorporates a new category not included in the training set, what value/index should be assigned? My approach would be assigning index=-1.

I was planning to implement a scorecard class in a few months, but we could start now. I have some previous experience developing scorecards, but it would be great if you could share your ideas. Feel free to provide an example with some data to start sketching the class! 

Thanks!

**Comment by guillermo-navas-palencia** [comment]
I haven't written unit tests yet, but you can try the ``metric="indices"`` in this branch: https://github.com/guillermo-navas-palencia/optbinning/tree/transform_indices

**Comment by GabrielSGoncalves** [comment]
Hi Guilhermo,

I'll test it and give you the feedback by tomorrow.
I'm also organizing the logic I'm using to develop a scorecard to share with you.
It's still a draft, but we could work together to find the best way to do it.

Thanks, 
Gabriel

**Comment by GabrielSGoncalves** [comment]
Hi Guilhermo,

Getting the indices helped a lot, specially on the posterior steps for doing the one-hot encoding and organizing the features categories.

I think this method would be complete if it also offered the possibility to get the bin names.
For example:
```python
from optbinning import OptimalBinning
from sklearn.datasets import load_breast_cancer

data = load_breast_cancer()
df = pd.DataFrame(data.data, columns=data.feature_names)
variable = "mean radius"
x = df[variable].values
y = data.target
optb = OptimalBinning(name=variable, dtype="numerical", solver="cp")
optb.fit(x, y)
```
Using the `transform` method I would get:
```python
optb.transform(x, metric="indices")[:20]
>> array([6, 6, 6, 0, 6, 2, 6, 4, 2, 2, 5, 5, 6, 5, 4, 4, 4, 5, 6, 3])
```
What would be extremely handy:
```python
optb.transform(x, metric="bins")[:20]
>> ['[16.93, inf)',
 '[16.93, inf)',
 '[16.93, inf)',
 '[-inf, 11.43)',
 '[16.93, inf)',
 '[12.33, 13.09)',
 '[16.93, inf)',
 '[13.70, 15.05)',
 '[12.33, 13.09)',
 '[12.33, 13.09)',
 '[15.05, 16.93)',
 '[15.05, 16.93)',
 '[16.93, inf)',
 '[15.05, 16.93)',
 '[13.70, 15.05)',
 '[13.70, 15.05)',
 '[13.70, 15.05)',
 '[15.05, 16.93)',
 '[16.93, inf)',
 '[13.09, 13.70)']
```
The way I'm getting this information is by using a list comprehension:
```python
[    optb.binning_table.build().at[x, "Bin"]
    for x in optb.transform(x, metric="indices")[:20] ]
```

I'm working on the design for a credit scoring class and I'm planning to share with you by the end of the day.

Att

Gabriel

**Comment by guillermo-navas-palencia** [comment]
Hi Gabriel,

I liked the ``metric="bins"``, the last commit includes the implementation of this option for all optimal binning algorithms in the library.

Guillermo

**Comment by guillermo-navas-palencia** [comment]
I just added a new issue for scorecard #27

**Comment by GabrielSGoncalves** [comment]
Hi Guillermo,

I tested the new feature, `transform(metric="bins")` and it works flawlessly.
I think the class `OptimalBinning` has everything needed for developing a robust scorecard framework.

I'm going to start using the #27 for our discussion over the subject.

Thanks a lot!

Gabriel
